wit/types.wit

```diff
@@ -26,14 +26,60 @@ interface types {
     other(string)
   }
 
-  // TODO: perhaps better align with HTTP semantics?
-  // This type enumerates the different kinds of errors that may occur when
-  // initially returning a response.
+  // The cases of this variant correspond to the IANA HTTP Proxy Error Types:
+  //   https://www.iana.org/assignments/http-proxy-status/http-proxy-status.xhtml#table-http-proxy-error-types
+  // Note: the `http_request_error` case maps to an `ok` response with a 4xx
+  // `status-code`.
```

Comment:
How exactly would that happen? Do we expect hosts to generate fake responses instead of returning `error`?

Reply:
(Depending on how recent discussion in #49 resolves, it sounds like this comment might just go away and we'll add an `http-request-error` instead.)

Reply:
... and that's what I did in [this](https://github.com/WebAssembly/wasi-http/pull/52/commits/d25c0ab39b9b1818a414337c0ce8a18c3e7b5553) last commit.  @acfoltzer make sense to you too?

Reply:
I don't think returning e.g. `http-request-error(413)` makes for a great developer experience when host rejects outgoing HTTP request because of size limits.

Reply:
What do you think should happen in that case?

Reply:
I think we need to add more variants to cover the aspects of an `outgoing-request` that the handler implementation might want or have to reject. The `proxy-status` RFC is not concerned with things like request URI size limits, because the proxy can just return a 414 directly to the client, so we shouldn't expect that it'll comprehensively cover the cases we're concerned with.

If we sufficiently cover the reasons that an embedder might reject an `outgoing-request` through variants in this error, enforcement when constructing the `outgoing-request`, setting the request's headers, or writing to its body, then I don't think it makes sense to include `http-request-error` for the reasons @PiotrSikora mentions.

@lukewagner, we've recently discussed a slice of this in our discussions of header filtering—this is one reason the embedder might want to reject a request. We should consider adding direct variants that would correspond to use of `HTTP-request-error` with 4xxs that speak to the other potential reasons an embedder may need to reject:

- `HTTP-request-length-required` (corresponds to 411)
- `HTTP-request-content-too-large` (corresponds to 413)
- `HTTP-request-URI-too-long` (corresponds to 414)
- `HTTP-response-header-section-size(u32)` and `HTTP-response-header-size(field-size-payload)` and trailer variants (corresponds to 431)

Some of the other 4xx codes that `proxy-status` contemplates for `HTTP-request-error` include codes like 403 that I think deserve a more specific reason if the embedder would be tempted to use it (like the presence of an illegal header), or codes like 405 which speak to properties that only the origin can know.